ユーザーテーブルの実装

ogadra opened the pull request:
ユーザーテーブルを実装することにより、データベース内でLINEの友達を管理出来るようにする

## やったこと
- 友達登録時にuser_idを取得
- user_id及び登録日時をDBに登録
- 友達削除時にDBから削除

## なぜしたのか
- 将来クーポン情報とLINEのユーザーテーブルを紐付けられるように、ユーザーのテーブルが欲しかった

## どうやったのか
- ユーザーテーブルを追加した
- Followイベントがあった場合に、`event['source']['userId']`と`event['timestamp']`を取得し、Userテーブルに追加するようにした
- Unfollowイベントがあった場合に、userIdの一致するデータを削除するようにした

## どうなるのか
- Followイベントを受け取ったとき(LINE BOTがユーザーに追加されたとき、及びブロック解除されたとき)にデータベースにuser_idとtimestamp情報が追加されるようにした
- Unfollowイベントを受け取ったとき(LINE BOTがユーザーにブロックされたとき)にデータベースからユーザーが削除されるようにした

ogadra committed:
友達登録時にuserIdを返すメソッドを追加

ogadra committed:
add user table

ogadra committed:
unfollow時にUserテーブルから削除

akkt222 commented on app/controllers/user_controller.rb:
```diff
@@ -0,0 +1,2 @@
+class UserController < ApplicationController
+end
```
[nits] このファイル使わないのであれば消しておいてください

ogadra replied:
Ruby on Rails初心者で恐縮なのですが、どのファイルが不必要でどのファイルが必要なファイルなのか、ということがまだよく分かっていません…
これはディレクトリから削除してしまって問題ないということでしょうか

akkt222 replied:
ですです
今の状態だとこのコントローラーを呼び出すことはないし、何も定義されていないので使う時に追加する方が良いと思います！
理由としては不要なコードをデプロイしてもエラーになり得てもそもそも使わないのでプラスになる要素がないであるとか、git的に差分は出来るだけ少ない方が見やすいとかですね

akkt222 commented on app/controllers/webhook_controller.rb:
```diff
@@ -34,6 +34,18 @@ def callback
+        User.create(user_id: event['source']['userId'], timestamp: timestamp_datetime)
+
+      when Line::Bot::Event::Unfollow
+        User.find_by(user_id: event['source']['userId']).destroy
```
[imo] ここは2行に分かれていた方が読みやすくて個人的には好きです
こういうイメージです
```
user = User.find_by(user_id: event['source']['userId'])
user.destroy
```

akkt222 replied:
[nits] rubyでは`destory`と`destory!`では挙動が異なり、前者は失敗した場合にfalseを返し後者は例外を挙げます。
考慮した上でもし`destory`を採用したのであれば良いのですが、そういう訳ではないのであればどちらの方が嬉しいか一考してもらえると🙇🏻‍♂️

akkt222 replied:
現状はまだ考慮いらないかもですが、Userを消した時にユーザーが持っているCouponもDBから消すのかは考えておいても良さそうですね🤔

akkt222 commented on app/models/user.rb:
```diff
@@ -0,0 +1,4 @@
+class User < ApplicationRecord
+  validates :user_id, presence: true, uniqueness: true
+  validates :timestamp, presence: true
+end
```
[nits] ファイルの末尾には改行を入れて欲しいです！
エディタの設定で常にそうなるようにしちゃえば意識する必要もなくなります。

akkt222 commented on db/migrate/20210818023159_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
+      t.string :user_id
```
[imo] この`user_id`という命名ですが、ここに入るのはLINEのIDだと思うので、それが分かるような名付けでも良いかなあという気がします。

akkt222 commented on db/migrate/20210818023159_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :users do |t|
+      t.string :user_id
+      t.datetime :timestamp
```
[ask] `timestamp`というカラムを定義してもらったと思うのですが、Userの作成時に現在日時を入れる「だけ」であれば元々ある`created_at`で良い気もするんですが、何か意図ってあったりします？？
あと、nullable(nullになり得る)カラムとして定義されていそうなので、特別な意図がなければ`null: false`をつけてあげた方が良いかも👀

RyotaNakaya replied:
同様に、`user_id` にはユニーク制約を付与した方が良さそう。

ogadra replied:
LINEの友達登録された時間とデータベースに入る時間は異なるので、何らかの事情で友達登録がなされた時間とデータベースに登録される時間に大きな乖離が起きる可能性を考慮して作りました。
`null: false`は付けて良さそうです。

akkt222 commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,7 @@
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
ditto

akkt222 commented on test/fixtures/users.yml:
```diff
@@ -0,0 +1,9 @@
+
+two:
+  user_id: MyString
+  timestamp: 2021-08-18 11:31:59
```
このファイルももし使わないなら要らないかも？

akkt222 reviewed (commented):
コメントしました！

- 今回はPRのdescriptionに**やったこと**(= what)を書いてもらったと思うのですが、他に**なぜやったのか**(= why)と**どうやったのか**(= how)は今後書いてもらうと良いかなと思います！
  - 特にwhatとhowは分離していると良さそうです

- レビュワーにはこの辺を重点的に見て欲しいよ、っていう観点があればそれも書いてあるとなお良しです
- gitのブランチには意味のある名付けをするのが良いです。今回なら自分ならfeature/create-users-tableとかにします
-  このPRの結果こういう動作になります、というような補足があるとレビューしやすいです
  - ついでに書ける場合は手元でどうするとその挙動を再現するのか書いてもらえるとなお良いです

RyotaNakaya commented on app/controllers/webhook_controller.rb:
```diff
@@ -34,6 +34,18 @@ def callback
+          type: 'text',
+          text: event['source']['userId']
+        }
+        client.reply_message(event['replyToken'], message)
```
[IMO]このメソッドは永続化処理の後に呼んだ方が良さそうですね。例外処理をどれだけリッチにするかにも依存するんですが、何らかの理由で登録に失敗した場合にメッセージなどを変えたい場合があると思います。

RyotaNakaya commented on db/schema.rb:
```diff
@@ -10,11 +10,18 @@
 
+  create_table "users", force: :cascade do |t|
+    t.string "user_id"
+    t.datetime "timestamp"
```
[ask] この `timestamp` カラムは何用のカラムですか？というのと `timestamp` だと何を表すのかわからないのでもう少し説明的な名前にした方が良いですね。

ogadra committed:
delete some files
deleted:    app/controllers/user_controller.rb
deleted:    test/fixtures/users.yml
deleted:    test/models/user_test.rb

ogadra committed:
ファイル末尾に改行を追加
modified:   app/models/user.rb

ogadra committed:
rename columns

ogadra committed:
友達登録時のメッセージ送信をDB登録後に変更

ogadra committed:
DBを作り直して制約を追加

ogadra committed:
ユーザーIDにユニーク制約を追加
ユーザーIDが重複して起こるバグを防ぐため

RyotaNakaya reviewed (approved):
![LGTM](https://image.lgtmoon.dev/74465)

akkt222 reviewed (approved):
![LGTM](https://image.lgtmoon.dev/137379)
